src/components/memory/MemoryCtrl.ts

```diff
@@ -0,0 +1,127 @@
+  memoryTypes: {value: string}[];
+
+  fields: string[] = [];
+  allfields: string [] = [];
+  prevData: any;
+  type: string = "";
+
+  constructor() {
+    this.fields.length = 0;
+    this.memoryTypes = MEMORY_TYPES;
+    if (isNotNullOrUndefined(this.configCallbacks)) {
+      this.configCallbacks.push(this);
+    }
+    if (isNotNullOrUndefined(this.data)) {
+      for (var field in this.data) {
+        this.setMemoryType(field, field);
+      }
+    }
+
+    this.prevData = {};
+    this.cleanMetadata();
+  }
+
+  setMemoryType(oldType: string, newType: string): void {
+    if (isNotNullOrUndefined(newType)) {
+      this.fields.length = 0;
+      this.type = newType;
+      this.memoryTypeData = isNotNullOrUndefined(this.data[oldType]) ? this.data[oldType] : {};
+      this.memoryTypeMeta = angular.copy(deepGet(this.meta, "children.memory.model-description." + newType + ".attributes"));
+      for (var field in this.memoryTypeMeta) {
+        this.fields.push(field);
+      }
+      if (isNotNullOrUndefined(oldType) && oldType !== newType) {
+        delete this.data[oldType];
+        this.memoryTypeData.removeType = oldType;
+      }
+      this.data[newType] = this.memoryTypeData;
+      this.memoryTypeData.type = newType;
+      this.memoryTypeMeta.type = this.makeMetaForType();
+
+      this.allfields = angular.copy(this.fields);
+      this.allfields.push("type");
+    }
+  }
+
+  changeMemoryType: Function = () => {
+    if (this.type !== this.memoryTypeData.type) {
+      this.setMemoryType(this.type, this.memoryTypeData.type);
+      makeFieldDirty(this.memoryTypeMeta.type);
+    }
+  };
+
+  undoMemoryTypeChange: Function = () => {
+    this.undoFieldChange("type");
+    this.setMemoryType(this.type, this.prevData.type);
+  };
+
+  isMemoryDefined(): boolean {
+    return isNotNullOrUndefined(this.data);
+  }
+
+  createDefaultMemory(): void {
+    this.data = {};
+    this.data.OBJECT = {size: "1000"};
+    this.setMemoryType("OBJECT", "OBJECT");
+    makeFieldDirty(this.memoryTypeMeta.size);
+    this.data["is-new-node"] = true;
+  }
+
+  isAnyFieldModified(): boolean {
+    return this.allfields.some(attrName => isFieldValueModified(this.memoryTypeMeta[attrName]));
+  }
+
+  isRestartRequired(): boolean {
+    return this.allfields.some(attrName => isFieldValueModified(this.memoryTypeMeta[attrName])
+    && fieldChangeRequiresRestart(this.memoryTypeMeta[attrName]));
+  }
+
+  cleanMetadata(): void {
+    this.allfields.forEach(attrName => {
+      makeFieldClean(this.memoryTypeMeta[attrName]);
+      this.prevData[attrName] = angular.copy(this.memoryTypeData[attrName]);
+    });
+  }
+
+  getStyle(field: string): string {
+    return isNotNullOrUndefined(this.memoryTypeMeta[field]) ? this.memoryTypeMeta[field].style : "";
+  }
+
+  private undoFieldChange(field: string): void {
+    this.memoryTypeData[field] = this.prevData[field];
+    makeFieldClean(this.memoryTypeMeta[field]);
+  }
+
+  private makeMetaForType(): any {
```

Comment:
Do we really need this method? It's only called once and it only returns a one liner